lib/rules/no-template-children.js

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+'use strict';
+
+const { docUrl } = require('../util/doc-url');
+
+const disallowedProperties = new Set([
+    'children',
+    'childNodes',
+    'firstChild',
+    'firstElementChild',
+    'lastChild',
+    'lastElementChild',
+]);
+
+/**
+ * Returns true if this is a `this.template` expression
+ */
+function isThisDotTemplateExpression(node) {
+    return (
+        node.type === 'MemberExpression' &&
+        node.object.type === 'ThisExpression' &&
+        node.property.type === 'Identifier' &&
+        node.property.name === 'template'
+    );
+}
+
+/**
+ * Returns true if this is an ObjectExpression property that should be disallowed for `this.template`
+ * E.g.: `const { firstChild } = this.template`
+ */
+function propertyIsDisallowedForThisDotTemplate(prop) {
+    return prop.key.type === 'Identifier' && disallowedProperties.has(prop.key.name);
+}
+
+/**
+ * Returns true if this is an ObjectExpression property that should be disallowed for `this`
+ * E.g.: `const { template: { firstChild } } = this`
+ */
+function propertyIsDisallowedForThis(prop) {
+    const { key, value } = prop;
+    return (
+        key.type === 'Identifier' &&
+        key.name === 'template' &&
+        value.type === 'ObjectPattern' &&
+        value.properties.some(propertyIsDisallowedForThisDotTemplate)
+    );
+}
+
+module.exports = {
+    meta: {
+        docs: {
+            description: 'prevent accessing the immediate children of this.template',
+            category: 'LWC',
+            recommended: true,
+            url: docUrl('no-template-children'),
+        },
+
+        schema: [],
+    },
+
+    create(context) {
+        return {
+            MemberExpression(node) {
+                const { object, property } = node;
+
+                if (
+                    isThisDotTemplateExpression(object) &&
+                    property.type === 'Identifier' &&
+                    disallowedProperties.has(property.name)
+                ) {
+                    // E.g.: `const child = this.template.firstChild`
+
+                    context.report({
+                        message: `Accessing ${property.name} on this.template is unsafe`,
+                        node: property,
+                    });
+                }
+            },
+            VariableDeclarator(node) {
+                const { id, init } = node;
+
+                if (
+                    id.type === 'ObjectPattern' &&
+                    isThisDotTemplateExpression(init) &&
```

Comment:
My first observation looking at this check is that `init` is optional and therefore calling `isThisDotTemplateExpression` before checking if it isn't `null` will throw. However, it is never the case because if the identifier is an `ObjectPattern`, the variable declaration has to be initialized.

```
let { children };  // Parsing error
```

If we were to swap out lines 88 with 89, it would fail. I would recommend checking if `init` is not `null` to protect ourselves against unexpected issues if we were to refactor this code in the future.

Reply:
Done.